community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledOrderabilityUtils.java

```diff
@@ -0,0 +1,339 @@
+                 rhs instanceof Comparable )
+            {
+                return ((Comparable) lhs).compareTo( rhs );
+            }
+
+            throw new IncomparableValuesException( lhs.getClass().getSimpleName(), rhs.getClass().getSimpleName() );
+        }
+    };
+
+    private static Comparator NUMBER_COMPARATOR = new Comparator<Number>()
+    {
+        @Override
+        public int compare( Number lhs, Number rhs )
+        {
+            //if floats compare float values if integer types, compare long values
+            if ( lhs instanceof Double && rhs instanceof Float )
+            {
+                return ((Double) lhs).compareTo( rhs.doubleValue() );
+            }
+            else if ( lhs instanceof Float && rhs instanceof Double )
+            {
+                return -((Double) rhs).compareTo( lhs.doubleValue() );
+            }
+            else if ( lhs instanceof Float && rhs instanceof Float )
+            {
+                return ((Float) lhs).compareTo( (Float) rhs );
+            }
+            else if ( lhs instanceof Double && rhs instanceof Double )
+            {
+                return ((Double) lhs).compareTo( (Double) rhs );
+            }
+            // Right hand side is neither Float nor Double
+            else if ( (lhs instanceof Double || lhs instanceof Float) )
+            {
+                return MathUtil.compareDoubleAgainstLong( lhs.doubleValue(), rhs.longValue() );
+            }
+            // Left hand side is neither Float nor Double
+            else if ( (rhs instanceof Double || rhs instanceof Float) )
+            {
+                return -MathUtil.compareDoubleAgainstLong( rhs.doubleValue(), lhs.longValue() );
+            }
+            //everything else is a long from cyphers point-of-view
+            return Long.compare( lhs.longValue(), rhs.longValue() );
+        }
+    };
+
+    private static Comparator STRING_COMPARATOR = new Comparator<Object>()
+    {
+        @Override
+        public int compare( Object lhs, Object rhs )
+        {
+            if ( lhs instanceof Character && rhs instanceof String )
+            {
+                return lhs.toString().compareTo( (String) rhs );
+            }
+            else if ( lhs instanceof String && rhs instanceof Character )
+            {
+                return ((String) lhs).compareTo( rhs.toString() );
+            }
+            else
+            { return ((Comparable) lhs).compareTo( rhs ); }
+        }
+    };
+
+    private static Comparator BOOLEAN_COMPARATOR = new Comparator<Boolean>()
+    {
+        @Override
+        public int compare( Boolean lhs, Boolean rhs )
+        {
+            return lhs.compareTo( rhs );
+        }
+    };
+
+    private static Comparator LIST_COMPARATOR = new Comparator<Object>()
+    {
+        @Override
+        public int compare( Object lhs, Object rhs )
+        {
+            Iterator lhsIter = (lhs.getClass().isArray()) ? arrayToIterator( lhs ) : ((List) lhs).iterator();
```

Comment:
Yes, these are private comparators and the public entry point to this class (the static `compare` method) handles `null` at the top.
The definition of the `LIST` super type, which uses this comparator is `value instanceof List<?> || value.getClass().isArray()`.